docs/EnvironmentBanner.md
# Environment Banner

## Introduction

In keeping with [SSDR policy][confluence-env-banners], an environment banner
should be displayed when running the application locally, or in the
sandbox, test, or qa Kubernetes environments, indicating the environment.

## Environment Banner Settings

This application uses the "[django-admin-notice][django-admin-notice]" package
for configuring and displaying the environment banner.

The environment banner is configured via environment variables, which are used
to provide the settings in the "src/umd_handle/settings.py" file.

The "django-admin-notice" package uses different settings than the environment
variables typically used in SSDR applications. The following table maps the
environment variables to the "django-admin-notice" settings:

| Environment variable          | django-admin-notice setting |
| ----------------------------- | --------------------------- |
| ENVIRONMENT_BANNER            | ADMIN_NOTICE_TEXT           |
| ENVIRONMENT_BANNER_FOREGROUND | ADMIN_NOTICE_TEXT_COLOR.    |
| ENVIRONMENT_BANNER_BACKGROUND | ADMIN_NOTICE_BACKGROUND.    |

* `ENVIRONMENT_BANNER` - the text to display in the banner
* `ENVIRONMENT_BANNER_FOREGROUND` - CSS color code (i.e., '#fff') for the
   foreground text
* `ENVIRONMENT_BANNER_BACKGROUND` - CSS color code (i.e., '#008000') for the
  background text

When the Django `DEBUG` setting is `True`, the "Local Environment" banner will
be displayed by default (as configured in the "settings.py" file).

When the Django `DEBUG` setting is `False`, the banner will not be displayed
`ENVIRONMENT_BANNER` environment variable is not set of empty.

---
[confluence-env-banners]: https://umd-dit.atlassian.net/wiki/spaces/LIB/pages/21431530/Create+Environment+Banners
[django-admin-notice]: https://github.com/DoctorJohn/django-admin-notice


docs/DevelopmentEnvironmentLocal.md
# Development Environment - Local

## Introduction

This document provides guidance on setting up a umd-handle Django development
environment on a local workstation.

## Prerequisites

* Python 3.14

* Install `libxmlsec1`. This is required for the SAML authentication using
  [djangosaml2].

  On Mac, it is available via Homebrew:

  ```zsh
  brew install xmlsec1
  ```

  On Debian or Ubuntu Linux, it is available via `apt`:

  ```zsh
  sudo apt-get install xmlsec1
  ```

* Update the `/etc/hosts` file to add:

  ```zsh
  127.0.0.1 handle-local
  ```

## Application Setup

1) Clone umd-handle-django from GitHub:

    ```zsh
    git clone git@github.com:umd-lib/umd-handle-django.git
    cd umd-handle-django
    ```

2) Set up the Python virtual environment, and install the dependencies

    ```zsh
    python -m venv .venv
    source .venv/bin/activate
    pip install -e '.[test]'
    ```

3) Download the `handle-test-sp.key` and `handle-test-sp.crt` files from the
   "handle-test-saml" entry in LastPass into the project rooo directory.

    ℹ️ Note: These files can be placed in an directory outside the project,
    if desired. Also, when downloading the `handle-test-sp.crt`
    file, Google Chrome will modify the file extension, by default, to ".cer".
    Be sure to to specify the ".crt" extension when downloading the file.
    Mozilla Firefox preserves the ".crt" extension.

4) Copy the "env_example" file to ".env":

    ```zsh
    cp env_example .env
    ```

    and update the following variables with the appropriate values:

    * SAML_KEY_FILE - relative (or absolute) file path to the
      `handle-test-sp.key` file
    * SAML_CERT_FILE - relative (or absolute) file path to the
      ``handle-test-sp.crt` file
    * SECRET_KEY - Either comment out (a random key will be automatically
      generated), or populate with anything with sufficient randomness,
      i.e. `uuidgen | shasum -a 256 | cut -c-64`
    * XMLSEC1_PATH - The full file path to the "xmlsec1" binary, (usually
      findable by running `which xmlsec1`)

5) Initialize the database:

    ```zsh
    ./src/manage.py migrate
    ```

6) Create an administrative user:

   ```zsh
   ./src/manage.py createsuperuser
   ```

7) Run the application. The default port is 3000; this is also the port that
   is registered with DIT to allow SAML authentication to work from local:

    ```zsh
    ./src/manage.py runserver
    ```

    The application will be running at <http://handle-local:3000/>

    ----

    ℹ️ **Note:** When running locally with Rails applications (such as Archelon)
    that also run on port 3000, the port should be changed using an
    argument, i.e.:

    ```zsh
    ./src/manage.py runserver 3001
    ```

    When running on a different port the admin GUI will not be accessible,
    as it will not be possible to login via CAS. The backend API (which uses
    JWT tokens), will be available.

    ----

    The administrative interface is accessible at
    <http://handle-local:3000/admin>.

## Tests

This project uses [pytest] in conjunction with the [pytest-django] plugin
to run its tests. To run the test suite:

```zsh
pytest
```

To run with coverage information:

```zsh
pytest --cov src --cov-report term-missing
```

## Linter

This project uses [ruff] to lint and format the code.

To run the linter:

```zsh
ruff check
```

[djangosaml2]: https://djangosaml2.readthedocs.io/
[pytest]: https://pytest.org/
[pytest-django]: https://pytest-django.readthedocs.io/en/latest/
[ruff]: https://github.com/astral-sh/ruff

src/umd_handle/settings.py
"""
Django settings for umd_handle project.

Generated by 'django-admin startproject' using Django 5.2.7.

For more information on this file, see
https://docs.djangoproject.com/en/5.2/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/5.2/ref/settings/
"""

from csp.constants import SELF
from django.core.management.commands.runserver import Command as runserver
from django.core.management.utils import get_random_secret_key
from environ import Env
from pathlib import Path
import saml2.saml
import saml2.xmldsig
from socket import gethostname, gethostbyname
from urlobject import URLObject

# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent.parent

# Static file dir
STATIC_ROOT = str(BASE_DIR / 'staticfiles')

# Take environment variables from .env file
Env.read_env(BASE_DIR / '.env')
env = Env()

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/5.2/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = env('SECRET_KEY', default=get_random_secret_key())

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = env.bool('DEBUG', False)

# JWT_SECRET: Secret key used to generate JWT tokens
JWT_SECRET = env('JWT_SECRET', default='')

SERVER_HOST = env.str('SERVER_HOST', '0.0.0.0')
SERVER_PORT = env.str('SERVER_PORT', '3000')
runserver.default_addr = SERVER_HOST
runserver.default_port = SERVER_PORT

BASE_URL = URLObject(env.str('BASE_URL', f"http://handle-local:{SERVER_PORT}/"))
CSRF_TRUSTED_ORIGINS = [str(BASE_URL.with_path(''))]

# Default list of allowed hosts
ALLOWED_HOSTS = [BASE_URL.hostname]

# Add the IP address (used by k8s health probes)
try:
    ALLOWED_HOSTS.append(gethostbyname(gethostname()))
except OSError:
    # ignore if we can't get the IP address
    pass

# Kubernetes internal host, used for inter-pod communication from the
# Handle.net server instance, and other Kubernetes pods
K8S_INTERNAL_HOST = URLObject(env.str('K8S_INTERNAL_HOST', 'umd-handle-app'))
ALLOWED_HOSTS.append(K8S_INTERNAL_HOST)

# The base URL (including a trailing slash) for the handle proxy server
# associated with this application. For example: "https://hdl-test.lib.umd.edu/"
HANDLE_HTTP_PROXY_BASE=URLObject(env.str('HANDLE_HTTP_PROXY_BASE', '<SET HANDLE_HTTP_PROXY_BASE>'))

# Application definition

INSTALLED_APPS = [
    'admin_notice',
    'django.contrib.admin',
    'django.contrib.auth',
    'umd_handle.api',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'django_extensions',
    'djangosaml2',
]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    "whitenoise.middleware.WhiteNoiseMiddleware",
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    'djangosaml2.middleware.SamlSessionMiddleware',
    'umd_handle.middleware.LoginRequiredMiddleware',
    'umd_handle.middleware.JWTAuthenticationMiddleware',
]

ROOT_URLCONF = 'umd_handle.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.request',
                'admin_notice.context_processors.notice',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'umd_handle.wsgi.application'


# Database
# https://docs.djangoproject.com/en/5.2/ref/settings/#databases
DATABASES = {
    'default': {
        'ENGINE': env.str('DB_ENGINE', 'django.db.backends.sqlite3'),
        'NAME': env.str('DB_NAME', BASE_DIR / 'db.sqlite3'),
        'USER': env.str('DB_USER', ''),
        'PASSWORD': env.str('DB_PASSWORD', ''),
        'HOST': env.str('DB_HOST', ''),
        'PORT': env.str('DB_PORT', ''),
    }
}

AUTHENTICATION_BACKENDS = (
    'django.contrib.auth.backends.ModelBackend',
    'umd_handle.auth.ModifiedSaml2Backend',
)

LOGIN_URL = '/saml2/login/'
LOGIN_URL_NAME = 'saml2_login' # A name for use in reverse() if needed
# Default to the Django admin page after login
LOGIN_REDIRECT_URL = '/admin'
SESSION_EXPIRE_AT_BROWSER_CLOSE = True

# Password validation
# https://docs.djangoproject.com/en/5.2/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]


# Internationalization
# https://docs.djangoproject.com/en/5.2/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/5.2/howto/static-files/

STATIC_URL = 'staticfiles/'

# Default primary key field type
# https://docs.djangoproject.com/en/5.2/ref/settings/#default-auto-field

DEFAULT_AUTO_FIELD = 'django.db.models.BigAutoField'

# SameSite Cookies
# The storage linked to it is accessible by default at request.saml_session.
SAML_SESSION_COOKIE_NAME = 'saml_session'
# By default, djangosaml2 will set “SameSite=None” for the SAML session cookie.
SAML_SESSION_COOKIE_SAMESITE = env('SAML_SESSION_COOKIE_SAMESITE', default='None')
# Remember that in your browser “SameSite=None” attribute MUST also have the “Secure” attribute,
# which is required in order to use “SameSite=None”, otherwise the cookie will be blocked.
SESSION_COOKIE_SECURE = env.bool('SESSION_COOKIE_SECURE', default=True)

# Handling Post-Login Redirects
SAML_ALLOWED_HOSTS = env('SAML_ALLOWED_HOSTS', cast=[str], default=ALLOWED_HOSTS)

SAML_DEFAULT_BINDING = saml2.BINDING_HTTP_POST
SAML_LOGOUT_REQUEST_PREFERRED_BINDING = saml2.BINDING_HTTP_POST
SAML_IGNORE_LOGOUT_ERRORS = True

# Users, attributes and account linking
SAML_CREATE_UNKNOWN_USER = True
SAML_ATTRIBUTE_MAPPING = {
    'uid': ('username',),
    'mail': ('email',),
    'givenName': ('first_name',),
    'urn:mace:umd.edu:sn': ('last_name',),
}
SAML_KEY_FILE = env('SAML_KEY_FILE', default='/etc/umd_handle/saml/key.pem')
SAML_CERT_FILE = env('SAML_CERT_FILE', default='/etc/umd_handle/saml/cert.pem')

SAML_CONFIG = {
    # full path to the xmlsec1 binary program
    'xmlsec_binary': env('XMLSEC1_PATH', default='/usr/bin/xmlsec1'),

    # your entity id, usually your subdomain plus the url to the metadata view
    'entityid': str(BASE_URL.netloc),

    # directory with attribute mapping
    'attribute_map_dir': str(BASE_DIR / 'attribute-maps'),

    # Permits to have attributes not configured in attribute-mappings
    # otherwise...without OID will be rejected
    'allow_unknown_attributes': True,

    # this block states what services we provide
    'service': {
        # we are just a lonely SP
        'sp': {
            'name': 'handle-local',
            'name_id_format': saml2.saml.NAMEID_FORMAT_TRANSIENT,
            # Define the authentication context
            'requested_authn_context': {
                'authn_context_class_ref': [
                    'urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport',
                    'urn:oasis:names:tc:SAML:2.0:ac:classes:TLSClient',
                ],
                'comparison': 'minimum',
            },

            # For Okta add signed logout requests. Enable this:
            # "logout_requests_signed": True,

            'endpoints': {
                # url and binding to the assertion consumer service view
                # do not change the binding or service name
                'assertion_consumer_service': [
                    # Path provided in the "AssertionConsumerService" tag in
                    # the service provider XML configuration provided to DIT
                    # when setting up the Rails "umd-handle" application.
                    (str(BASE_URL.with_path('/users/auth/saml/callback')), saml2.BINDING_HTTP_POST),
                ],
                # url and binding to the single logout service view
                # do not change the binding or service name
                'single_logout_service': [
                    # Disable next two lines for HTTP_REDIRECT for IDPs that only support HTTP_POST. Ex. Okta:
                    (str(BASE_URL.with_path('/saml2/ls/')), saml2.BINDING_HTTP_REDIRECT),
                    (str(BASE_URL.with_path('/saml2/ls/post/')), saml2.BINDING_HTTP_POST),
                ],
            },

            'signing_algorithm': saml2.xmldsig.SIG_RSA_SHA256,
            'digest_algorithm': saml2.xmldsig.DIGEST_SHA256,

            # Mandates that the identity provider MUST authenticate the
            # presenter directly rather than rely on a previous security context.
            'force_authn': False,

            # Enable AllowCreate in NameIDPolicy.
            'name_id_format_allow_create': False,

            # attributes that this project need to identify a user
            'required_attributes': ['givenName', 'sn', 'mail', 'eduPersonEntitlement'],

            # attributes that may be useful to have but not required
            'optional_attributes': [],

            'want_response_signed': False,
            'authn_requests_signed': True,
            'logout_requests_signed': True,
            # Indicates that Authentication Responses to this SP must
            # be signed. If set to True, the SP will not consume
            # any SAML Responses that are not signed.
            'want_assertions_signed': True,

            'only_use_keys_in_metadata': True,

            # When set to true, the SP will consume unsolicited SAML
            # Responses, i.e. SAML Responses for which it has not sent
            # a respective SAML Authentication Request.
            'allow_unsolicited': True,

            # in this section the list of IdPs we talk to are defined
            # This is not mandatory! All the IdP available in the metadata will be considered instead.
            'idp': {
                # we do not need a WAYF service since there is
                # only an IdP defined here. This IdP should be
                # present in our metadata

                # the keys of this dictionary are entity ids
                'https://shib.idm.umd.edu/shibboleth-idp/shibboleth': {
                    'single_sign_on_service': {
                        saml2.BINDING_HTTP_POST: 'https://shib.idm.umd.edu/shibboleth-idp/profile/SAML2/POST/SSO',
                    },
                    'single_logout_service': {
                        saml2.BINDING_HTTP_REDIRECT: 'https://shib.idm.umd.edu/shibboleth-idp/profile/Logout',
                    },
                },
            },
        },
    },

    # where the remote metadata is stored, local, remote or mdq server.
    # One metadata store or many ...
    'metadata': {
        'remote': [
            {'url': 'https://shib.idm.umd.edu/shibboleth-idp/shibboleth'},
        ],
    },

    # set to 1 to output debugging information
    'debug': 1,

    # Signing
    'key_file': SAML_KEY_FILE,
    'cert_file': SAML_CERT_FILE,

    # Encryption
    'encryption_keypairs': [{
        'key_file': SAML_KEY_FILE,
        'cert_file': SAML_CERT_FILE,
    }],
}

# Content security policy largely taken from
# https://django-csp.readthedocs.io/en/v4.0/configuration.html#configuration
CONTENT_SECURITY_POLICY = {
    "EXCLUDE_URL_PREFIXES": [],
    "DIRECTIVES": {
        "default-src": [SELF],
        "frame-ancestors": [SELF],
        "form-action": [SELF],
    },
}

# Environment banner setting - defaults to displaying local environment banner
# if DEBUG flag is set.
ADMIN_NOTICE_TEXT = env.str('ENVIRONMENT_BANNER', 'Local Environment' if DEBUG else '')
ADMIN_NOTICE_TEXT_COLOR = env.str('ENVIRONMENT_BANNER_FOREGROUND', '#fff')
ADMIN_NOTICE_BACKGROUND = env.str('ENVIRONMENT_BANNER_BACKGROUND', '#008000')
